digdag-core/src/main/java/io/digdag/core/plugin/RemotePluginLoader.java

```diff
@@ -0,0 +1,190 @@
+
+    private static RepositorySystemSession newRepositorySystemSession(RepositorySystem system, Path localRepositoryPath)
+    {
+        DefaultRepositorySystemSession session = MavenRepositorySystemUtils.newSession();
+
+        LocalRepository localRepo = new LocalRepository(localRepositoryPath.toString());
+        session.setLocalRepositoryManager(system.newLocalRepositoryManager(session, localRepo));
+
+        //session.setTransferListener(new ConsoleTransferListener());
+        //session.setRepositoryListener(new ConsoleRepositoryListener());
+
+        return session;
+    }
+
+    private final RepositorySystem system;
+    private final RepositorySystemSession session;
+
+    public RemotePluginLoader(Path localRepositoryPath)
+    {
+        this.system = newRepositorySystem();
+        this.session = newRepositorySystemSession(system, localRepositoryPath);
+    }
+
+    @Override
+    public PluginSetFactory load(Spec spec)
+    {
+        if (spec.getDependencies().isEmpty()) {
+            return PluginSetFactory.empty();
+        }
+
+        ImmutableList.Builder<PluginFactory> builder = ImmutableList.builder();
+
+        List<RemoteRepository> repositories = getRepositories(spec);
+
+        for (String dep : spec.getDependencies()) {
+            // download artifacts, and/or resolve local-repository references to them
+            logger.debug("Loading plugin {}", dep);
+            List<ArtifactResult> artifactResults = resolveArtifacts(repositories, dep);
+
+            logger.debug("Classpath of plugin {}: {}", dep,
+                    artifactResults.stream().map(a -> a.getArtifact().getFile().toString())
+                    .collect(Collectors.joining(File.pathSeparator)));
+
+            ClassLoader pluginClassLoader = buildPluginClassLoader(artifactResults);
+            try {
+                ServiceLoader<PluginFactory> serviceLoader = ServiceLoader.load(PluginFactory.class, pluginClassLoader);
+                for (PluginFactory factory : serviceLoader) {
+                    builder.add(factory);
+                }
+            }
+            catch (ServiceConfigurationError ex) {
+                throw new RuntimeException("Failed to lookup io.digdag.spi.PluginFactory service from a dependency '" + dep + "'", ex);
+            }
+        }
+
+        return new PluginSetFactory(builder.build());
+    }
+
+    private ClassLoader buildPluginClassLoader(List<ArtifactResult> artifactResults)
+    {
+        ImmutableList.Builder<URL> urls = ImmutableList.builder();
+        for (ArtifactResult artifactResult : artifactResults) {
+            URL url;
+            try {
+                url = artifactResult.getArtifact().getFile().toPath().toUri().toURL();
+            }
+            catch (MalformedURLException ex) {
+                throw Throwables.propagate(ex);
+            }
+            urls.add(url);
+        }
+        return new PluginClassLoader(urls.build(), RemotePluginLoader.class.getClassLoader(),
```

Comment:
`io.digdag.core.plugin.PluginClassLoader` works as following:

* If a class is in shared packages (`io.digdag.spi`), lookup from the parent (system classloader) first, then lookup from the child (classpath of the plugin).
* otherwise, opposite order: lookup from the child (classpath of the plugin), then lookup from the parent (system classloader)

(this idea is taken from Presto's [PluginClassLoader](https://github.com/prestodb/presto/blob/c7bf63a2138c2e61f0edb307cefbf6b35fba25c2/presto-main/src/main/java/com/facebook/presto/server/PluginManager.java#L295-L299), by the way).

Dependency conflicts don't happen with this way because all classes excepting io.digdag.spi are loaded first from the plugin's classpath.

Difference is that if a plugin includes missing classes (which is probably not a big deal), my idea tires to load them from the parent classloader, and your idea throws NoClassDefFoundError immediately.